### PR TITLE
Implement clock() and sleep family of functions

### PR DESCRIPTION
**Pull request by runer112**
The clock uses a single 32768Hz 32-bit timer. It will overflow after ~36 hours, but that's probably acceptable for a calculator program.

Sleep functions added:
1. `msleep()`: millisecond sleep, "alias" of `delay()`
2. `sleep()`: second sleep, implements https://man7.org/linux/man-pages/man3/sleep.3.html
3. `ticksleep()`: clock tick sleep
4. `usleep()`: microsecond sleep, implements https://man7.org/linux/man-pages/man3/usleep.3.html

### CODE REVIEW

**Comment by runer112** [comment]
Should address #321.

File: src/ce/crt0.src [hunk 1]
```diff
@@ -107,6 +107,47 @@ launcher:
 .webaddress:
 	db	"http://tiny.cc/clibs",0
 
+
+	section .start
+
+mpTmrRange    := 0F20000h
+
+tmr2Counter   := 10h
```
Comment (runer112) [inline src/ce/crt0.src:115]: Timer 2 is probably the best timer for this, right? I imagine that most programs with custom timer usage would use timer 1, and USB (including usbdrvce?) uses timer 3.

Comment (mateoconlechuga) [reply src/ce/crt0.src:115]: That sounds good to me

File: examples/standalone/stopwatch/readme.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ### Stopwatch Demo
 
 This demo demonstrates using the hardware timers to make a simple stopwatch
-displaying millisecond precision.
+displaying hundreths precision.
```
Comment (adriweb) [inline examples/standalone/stopwatch/readme.md:4]: `hundredths`
Also "of seconds"?

File: src/makefile.mk [hunk 3]
```diff
@@ -26,7 +26,7 @@ BSSHEAP_LOW ?= D052C6
 BSSHEAP_HIGH ?= D13FD8
 STACK_HIGH ?= D1A87E
 INIT_LOC ?= D1A87F
-OUTPUT_MAP ?= NO
+OUTPUT_MAP ?= YES
```
Comment (adriweb) [inline src/makefile.mk:29]: was that done on purpose here

Comment (mateoconlechuga) [reply src/makefile.mk:29]: I disagree with this modification

Comment (runer112) [reply src/makefile.mk:29]: Yes, this was done on purpose. I don't see why anyone wouldn't want this extra build artifact that makes debugging in CEmu not a total nightmare.

Mateo mentioned something about it being slow on Windows, but I tried building oiram with and without this and saw no noticeable change in build time.

Comment (mateoconlechuga) [reply src/makefile.mk:29]: Seems reasonable; I guess things have improved a lot

**Comment by runer112** [comment]
Now that there are `sleep()` and `usleep()`, `delay()` to sleep for a number of milliseconds feels a bit out of place. How about introducing an alias, like `msleep()`?

**Comment by mateoconlechuga** [comment]
> Now that there are `sleep()` and `usleep()`, `delay()` to sleep for a number of milliseconds feels a bit out of place. How about introducing an alias, like `msleep()`?

This makes sense to me.

File: examples/standalone/ticksleep/src/main.c [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+int main(void)
+{
+    /* Clear the homescreen */
+    os_ClrHome();
+
+    /* Wait for 5 seconds */
+    os_PutStrFull("Waiting...");
+    ticksleep(5 * CLOCKS_PER_SEC);
+    os_PutStrFull(" Done!");
+    delay(100);
```
Comment (mateoconlechuga) [inline examples/standalone/ticksleep/src/main.c:13]: why is there a delay here?

File: examples/standalone/usleep/src/main.c [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+int main(void)
+{
+    /* Clear the homescreen */
+    os_ClrHome();
+
+    /* Wait for 5 seconds */
+    os_PutStrFull("Waiting...");
+    usleep(5 * 1000000);
+    os_PutStrFull(" Done!");
+    delay(100);
```
Comment (mateoconlechuga) [inline examples/standalone/usleep/src/main.c:12]: and why a delay here... am I missing something?

File: src/ce/crt0.src [hunk 6]
```diff
@@ -105,6 +105,47 @@ launcher:
...
+tmrCtrl       := 30h
+
+bTmr2Enable   := 3
+bTmr2Crystal  := 4
+bTmr2Overflow := 5
+bTmr2CountUp  := 10
+
+	public	___init_clock
```
Comment (mateoconlechuga) [inline src/ce/crt0.src:121]: should this really be public? also it should be in a defined section

File: examples/standalone/ticksleep/src/main.c (hunk 4 above)
Comment (runer112) [reply examples/standalone/ticksleep/src/main.c:13]: That's just for the autotest to test the timing. It gives a 100ms target to check for after the expected time has elapsed.

File: src/ce/crt0.src (hunk 6 above)
Comment (mateoconlechuga) [reply src/ce/crt0.src:121]: is this code removed if the user isn't using clock() and related functions?

File: src/ce/delay.src [hunk 7]
```diff
@@ -1,90 +1,64 @@
...
+; Get millis.
 	pop	de
-	ex	(sp),hl			; 0hl = millis
+	ex	(sp), hl		; hl = millis
 	push	de
+; ^ 48 cc
 ; Abort ASAP if millis == 0.
-	ld	a,h
-	or	a,l
+	ld	a, l
```
Comment (mateoconlechuga) [inline src/ce/delay.src:20]: this is unsafe

File: src/ce/sleep.src [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+; Even though the input is 24-bit, let's just pretend that it's 16-bit.
+; Timing assumes fetching from RAM (4cc) and the first timer read being valid.
+
+; Get seconds.
+	pop	de
+	ex	(sp), hl		; hl = seconds
+	push	de
+; ^ 48 cc
+; Abort ASAP if seconds == 0.
+	ld	a, l
```
Comment (mateoconlechuga) [inline src/ce/sleep.src:21]: this is technically unsafe

File: src/ce/crt0.src (hunk 6 above)
Comment (runer112) [reply src/ce/crt0.src:121]: Yes, it needs to be public. Referencing this symbol (or not) is what determined whether the section it's in (and it is in one) will be included.

File: src/ce/usleep.src [hunk 9]
```diff
@@ -0,0 +1,60 @@
+; typedef unsigned int useconds_t;
+; int usleep(useconds_t usec);
+
+	assume	adl=1
+
+virtual
+	call	_usleep
```
Comment (mateoconlechuga) [inline src/ce/usleep.src:7]: why?

**Review by mateoconlechuga — CHANGES_REQUESTED** [review]

File: src/ce/crt0.src (hunk 6 above)
Comment (mateoconlechuga) [reply src/ce/crt0.src:121]: Oh that makes sense.

Comment (runer112) [reply src/ce/crt0.src:121]: I just double-checked. Yes, if none of the new timing functions are referenced, this initialization code is not included.

**Review by mateoconlechuga — APPROVED** [review]

**Review by mateoconlechuga — APPROVED** [review]

**Review by mateoconlechuga — APPROVED** [review]

**Pull request merged**